artifact/signer.go

```diff
@@ -158,16 +160,52 @@ func MarshalECDSASignature(r, s *big.Int) ([]byte, error) {
 }
 
 func UnmarshalECDSASignature(sig []byte) (r, s *big.Int, e error) {
-	// check if the size of the key matches provided one
-	if len(sig) != 2*ecdsa256keySize {
-		return nil, nil, errors.Errorf("signer: invalid ecdsa key size: %d", len(sig))
+	// if the length of the signature is the double of key size, we know (and assume)
+	// that it was us who created it via MarshalECDSASignature.
+	// After the introduction of the hardware security support there is another possibility
+	// (see below)
+	if len(sig) == 2*ecdsa256keySize {
+		// get the signature; see corresponding `Sign` function for more details
+		// about serialization
+		r = big.NewInt(0).SetBytes(sig[:ecdsa256keySize])
+		s = big.NewInt(0).SetBytes(sig[ecdsa256keySize:])
+		return r, s, nil
 	}
 
-	// get the signature; see corresponding `Sign` function for more details
-	// about serialization
-	r = big.NewInt(0).SetBytes(sig[:ecdsa256keySize])
-	s = big.NewInt(0).SetBytes(sig[ecdsa256keySize:])
-	return r, s, nil
+	// in case of a key supplied via PKCS#11 URI, we have no control over what the signature is
+	// since it is designed to be actually verified via the same mechanism (PKCS#11 URI).
+	// We know here that it is ECDSA key, and judging form the size we can assume
+	// that it is ASN.1 encoded. If so, then it should be 72 bytes, the less or equal is here
+	// to support keys of size 71 bytes (strangely seen in the wild). In other words:
+	// if the signature has not been created with MarshalECDSASignature, then we assume
+	// it is to be decoded via ASN.1, with the protection on the signature length.
+	if len(sig) <= ecdsa256Asn1SizeBytes {
+		var value asn1.RawValue
+		_, err := asn1.Unmarshal(sig, &value)
+		if err != nil {
+			return nil, nil, errors.Wrap(err, "cannot unmarshal asn1 data")
+		}
+		bytesValue := value.Bytes
+		if len(bytesValue) > 0 {
+			var v asn1.RawValue
+			bytesValue, err = asn1.Unmarshal(bytesValue, &v)
+			if err != nil {
+				return nil, nil, errors.Wrap(err, "cannot unmarshal asn1 r value")
+			}
+			r = big.NewInt(0).SetBytes(v.Bytes)
+		}
+		if len(bytesValue) > 0 {
+			var v asn1.RawValue
+			_, err = asn1.Unmarshal(bytesValue, &v)
+			if err != nil {
+				return nil, nil, errors.Wrap(err, "cannot unmarshal asn1 s value")
+			}
+			s = big.NewInt(0).SetBytes(v.Bytes)
+		}
+		return r, s, nil
+	}
+
+	return nil, nil, errors.Errorf("signer: invalid ecdsa key size: %d", len(sig))
```

Comment:
Please consider refactoring the whole thing.
```
func UnmarshalECDSASignature(sig []byte) (r, s *big.Int, e error) {
    if len(sig) == 2*ecdsa256keySize {
        return  UnmarshalECDSASignatureMenderArtifact(...)
    }
    if len(sig) <= ecdsa256Asn1SizeBytes {
        return  UnmarshalECDSASignatureASN1(...)
    }
    return error...
```

Reply:
Reasoning: it will make the code more readable and the comments contained inside separate functions